application/controllers/UserManagementController.php

```diff
@@ -1394,40 +1413,90 @@ protected function getRandomString(): string
-        $oCriteria = new CDbCriteria();
-        $oCriteria->compare('uid', $iUserId);
-        // without entity
-        $oCriteria->compare('entity_id', 0);
-        // except for template entity (no entity_id is set here)
-        $oCriteria->compare('entity', "<>template");
-        Permission::model()->deleteAll($oCriteria);
-
+        /**
+         * Get current user permission to update only this permission
+         * NEVER delete existing Permission !
+         */
+        $aAllowedPermissions = $this->getCurrentUserPermission();
         $results = [];
         //Apply the permission array
-        foreach ($aPermissionArray as $sPermissionKey => $aPermissionSettings) {
-            $oPermission = new Permission();
-            $oPermission->entity = 'global';
-            $oPermission->entity_id = 0;
-            $oPermission->uid = $iUserId;
-            $oPermission->permission = $sPermissionKey;
-
-            foreach ($aPermissionSettings as $sSettingKey => $sSettingValue) {
-                $oPermissionDBSettingKey = $sSettingKey . '_p';
-                $oPermission->$oPermissionDBSettingKey = $sSettingValue == 'on' ? 1 : 0;
+        foreach ($aAllowedPermissions as $permissionKey => $aAllowedPermission) {
+            /* get the current user permission or create */
+            $oPermission = Permission::model()->find(
+                "entity = :entity AND entity_id = :entity_id AND uid = :uid AND permission = :permission",
+                array(
+                    ":entity" => 'global',
+                    ":entity_id" => 0,
+                    ":uid" => $iUserId,
+                    ":permission" => $permissionKey,
+                )
+            );
+            if (empty($oPermission)) {
+                $oPermission = new Permission();
+                $oPermission->entity = 'global';
+                $oPermission->entity_id = 0;
+                $oPermission->uid = $iUserId;
+                $oPermission->permission = $permissionKey;
             }
-
-            $aPermissionData = Permission::getGlobalPermissionData($sPermissionKey);
-
-            $results[$sPermissionKey] = [
-                'descriptionData' => $aPermissionData,
+            foreach ($aAllowedPermission as $action => $havePermission) {
+                if ($havePermission) {
+                    $oPermission->setAttribute(
+                        $action . '_p',
+                        intval(!empty($aPermissionArray[$permissionKey][$action]))
+                    );
+                }
+            }
+            $results[$permissionKey] = [
+                'descriptionData' => Permission::getGlobalPermissionData($permissionKey),
                 'success' => $oPermission->save(),
                 'storedValue' => $oPermission->attributes
             ];
         }
         return $results;
     }
 
+    /**
+     * Get user permission (CRUD) as array
+     * @return array
+     */
+    private function getCurrentUserPermission(): array
```

Comment:
Should be in service class?

Reply:
No, we have it here : https://github.com/LimeSurvey/LimeSurvey/blob/9510ce9580e1f85cc3b1b980e5474413789c78b4/application/controllers/UserManagementController.php#L412

I have to apply it here too : same function twice. I make the commit.

About Service class : unsure it must be the same way for Object ? I check at same time.
(but prefer to have it merged monday)

Reply:
I move to a single funtcion for both part.

About in Service class : yes we have it , but need a complet GUI rewirite.

We use same Gui (near) for Survey and for SurveyGroup (and hope for user, Template etc …) but here : it's a security fix not a rewrite.

Reply:
I Survey we use disable : https://github.com/LimeSurvey/LimeSurvey/blob/9510ce9580e1f85cc3b1b980e5474413789c78b4/application/models/services/PermissionManager.php#L77 
Not hidden